include/aws/io/uri.h

```diff
@@ -0,0 +1,128 @@
+#ifndef AWS_IO_URI_H
+#define AWS_IO_URI_H
+/*
+ * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+#include <aws/io/io.h>
+
+#include <aws/common/byte_buf.h>
+
+/**
+ * Data representing a URI. uri_str is always allocated and filled in.
+ * The other portions are merely storing offsets into uri_str.
+ */
+struct aws_uri {
+    size_t self_size;
+    struct aws_allocator *allocator;
+    struct aws_byte_buf uri_str;
+    struct aws_byte_cursor scheme;
+    struct aws_byte_cursor authority;
+    struct aws_byte_cursor host_name;
+    uint16_t port;
+    struct aws_byte_cursor path;
+    struct aws_byte_cursor query_string;
+    struct aws_byte_cursor path_and_query;
+};
+
+/**
+ * key/value pairs for a query string. If the query fragment was not in format key=value, the fragment value
+ * will be stored in key
+ */
+struct aws_uri_param {
+    struct aws_byte_cursor key;
+    struct aws_byte_cursor value;
+};
+
+/**
+ * Arguments for building a URI instance. All members must
+ * be initialized before passing them to aws_uri_init().
+ */
+struct aws_uri_builder_options {
+    struct aws_byte_cursor scheme;
+    struct aws_byte_cursor path;
+    struct aws_byte_cursor host_name;
+    uint16_t port;
+    struct aws_array_list query_params;
```

Comment:
Since query strings don't HAVE to be key-value pairs, why not offer an alternative `aws_byte_cursor query_string` and the user can provide one or the other.